File: src/api/schemas/tool.py
"""Pydantic schemas for tools API (S5)."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class Pagination(BaseModel):
    page: int
    page_size: int
    total: int
    total_pages: int


# ── Tool registry (static metadata) ──────────────────────
class ToolParamInfo(BaseModel):
    name: str
    input_type: str
    flag: str
    required: bool = False
    multiple: bool = False


class ToolInfo(BaseModel):
    name: str
    binary: str
    description: str
    category: str
    params: list[ToolParamInfo] = []
    allowed_flags: list[str] = []
    default_timeout: int = 300
    enabled: bool = True


# ── Run tool ─────────────────────────────────────────────
class RunToolRequest(BaseModel):
    inputs: dict = Field(..., description="工具输入 {domain/ip/url: value | [values]}")
    flags: list[str] | None = None
    timeout: int | None = Field(default=None, ge=10, le=3600)
    task_id: UUID | None = None


class ToolExecutionResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: UUID
    tool_name: str
    category: str | None = None
    task_id: UUID | None = None
    inputs: dict
    flags: dict | None = None
    status: str
    exit_code: int | None = None
    stdout: str | None = None
    stderr: str | None = None
    parsed: list | dict | None = None
    duration_s: float | None = None
    error: str | None = None
    started_at: datetime
    created_at: datetime


class ToolExecutionListResponse(BaseModel):
    data: list[ToolExecutionResponse]
    pagination: Pagination

File: web/src/types/tool.ts
// S5 工具管理类型 — 对齐后端 routers/tools.py

export interface ToolParamInfo {
  name: string;
  input_type: string; // domain/ip/cidr/url/word
  flag: string;
  required: boolean;
  multiple: boolean;
}

export interface ToolInfo {
  name: string;
  binary: string;
  description: string;
  category: string; // recon/dns/portscan/vulnscan/cert
  params: ToolParamInfo[];
  allowed_flags: string[];
  default_timeout: number;
  enabled: boolean;
}

export interface RunToolRequest {
  inputs: Record<string, string | string[]>;
  flags?: string[];
  timeout?: number;
  task_id?: string;
}

export interface ToolExecution {
  id: string;
  tool_name: string;
  category: string | null;
  task_id: string | null;
  inputs: Record<string, unknown>;
  flags: string[] | null;
  status: string; // success/failed/timeout/unavailable/invalid_input
  exit_code: number | null;
  stdout: string | null;
  stderr: string | null;
  parsed: unknown;
  duration_s: number | null;
  error: string | null;
  started_at: string;
  created_at: string;
}

export interface Pagination {
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
}
